Add documentation on troubleshooting ACME (#21381)
hashicorp/vault: website/content/docs/secrets/pki/troubleshooting-acme.mdx

---
layout: docs
page_title: 'PKI - Secrets Engine: Troubleshooting ACME'

> description: The PKI secrets engine for Vault generates TLS certificates.

Author: This is also a consistent header for all pages -- should we take an action item to update this later, consistently, for all pages? 

See also: https://github.com/hashicorp/vault/pull/21381#discussion_r1237609561

> description: The PKI secrets engine for Vault generates TLS certificates.

Reviewer: I'm going to create tickets and PRs for that sort of maintenance next week, but I'd like us to start updating/correcting docs as we touch them as well. Partly to help socialize the style guide and partly because it's that kind of maintenance would be a monster PR if we don't handle it in pieces anyway.

---

# PKI Secrets Engine - Troubleshooting ACME

Below are suggestions for troubleshooting issues with the ACME clients
requesting access from Vault. Different clients will behave in different
ways and have different ways of viewing more details about failures.

> ## Contacting support

Author: Hmm, I wanted a very visible directive that says, hey, if you do contact support, be sure to include this extra info you aren't used to providing, because it makes finding the problem much easier.

> ## Contacting support

Reviewer: I would, yes. You can write this section so it has clear directives about what is required for proper support. The key to remember is that folks scan docs, they don't really read them. If they come to a troubleshooting page and the first item they see is "contact support", that's going to be where most readers stop.

> 

Author: People tend to contact support when they want to, I don't think we have much say in if they do/don't... :-) I don't know of any errors off the top of my head that would be best resolved by contacting support.

When contacting Hashicorp Support with ACME related issues or filing a
GitHub issue, the following pieces of information are helpful for Support
and Engineering to investigate and reproduce any issues related to ACME,

> in addition to the regularly asked pieces of information:

Author: I do not honestly know, tbh. :-) However, the remaining info is unique to ACME that I'd like to see included anyways.

I went looking briefly https://support.hashicorp.com/hc/en-us/requests/new doesn't really give any requirements, but I frequently see debug bundles and logs attached to reports. 

I suspect this isn't a standardized up-front requirement but instead depends on the support personnel to identify what they want... So having this as a template to refer to will be helpful for them.

> in addition to the regularly asked pieces of information:

Reviewer: Okay. For now then, I would recommend removing the phrase so we're not implying that there's some additional information they should somehow already know to include. I'll create a doc ticket to try and hunt that information down and create a partial we can use going forward.

 - **ACME client name and version**; different clients behave differently
   and certain ACME clients might have bugs or other shortcomings at

>    particular versions.

Author: Given we have justifications for others, I had this for consistency. I agree, its a standard requirement, but it is above and beyond what is (probably typically) negotiated during the regular support discussions... I dunno

>    particular versions.

Reviewer: TBH, I think I removed most of the other justifications in my edits 😅 At the end of the day, if there is information we require to handle support requests, we don't *really* need to justify the why of it, just make sure we explain how to get the information we need.

>    particular versions.

Author: Ah ok :-) That'd be fine by me!

>  - **ACME client logs**, if any; on failure, `certbot` will give the location

Author: certbot is not part of Vault, but it is one of the most popular clients and I'd really like to encourage users to grab these logs and submit them as part of requests as they're very helpful for debugging, more so than the command's stdout/stderr.

   of debug logs it has produced while attempting to request certificates.
   This frequently contains helpful information about requests to the Vault
   cluster and the server's response. Other clients may or may not behave
   similarly; in the event of not having log information, full output in

>    verbose (or debug mode) of the client would be helpful.

Author: I think on some clients, e.g., `acme.sh`, logs and output are the same (there is no separate logs). But maybe on others, there's only logs and no separate output... :shrug:  

That's mostly why I mixed the two together -- they're conceptually the same information, perhaps in two different places (in `certbot`'s case) or just one of them will exist (in the case of `acme.sh`). :-)

>    verbose (or debug mode) of the client would be helpful.

Reviewer: Yeah, I suspected that was the case :) My concern from a readability perspective is that it was a really dense paragraph and figuring out exactly what I need to do as a stressed practitioner might be difficult. If I'm not reading closely, I might come away with the idea that I have to provide both. That's why I suggest breaking them into two points and adding a caveat so it's really obvious that I can do one or the other but don't have to do both.

 - **Vault server DEBUG level logs**; many ACME errors are hidden from the
   client as they contain sensitive information, but are instead logged,
   when appropriate, to the server logs at a DEBUG level.

> ## Debugging cluster path issues

Author: This is not an enterprise-specific feature.

> ## Debugging cluster path issues

Reviewer: > This is not an enterprise-specific feature

Okay. The original text mentioned enterprise, so I thought that this was a problem that would only apply to Enterprise installations.

As for the headings, the main reason I'm recommending using the error message as the heading is SEO. When people run into an error message, one of the first troubleshooting steps is always to just paste the error text into a web search and see what comes up. Docs where a heading matches the string closely will rank higher in search results that docs where the string appears somewhere in the text.

> If ACME works on some nodes of a Vault Enterprise cluster but not on
> others, likely it means that the cluster address has not been set.

Reviewer: ```suggestion
```

I suggest moving the intro stuff to a "Cause" section after "Symptoms" so it is easier to see when scanning.

### Symptoms

When reading the ACME config on these Performance Secondary nodes, a
warning might appear like:

> ACME feature requires local cluster 'path' field configuration to be set

When attempting to contact such a node with an ACME client, a server internal
error will appear with the above message as well.

### Resolution

For each Performance Replication cluster, read the value of `/config/cluster`
and ensure the `path` field is set. When it is missing, update the URL to
point to this mount's path on a TLS-enabled address for this PR cluster; this
domain may be a load balanced or a DNS round robin address. For example:

```
$ vault write pki/config/cluster path=https://cluster-b.vault.example.com/v1/pki
```

Once this is done, re-read the ACME configuration and make sure no warnings
appear:

```
$ vault read pki/config/acme
```

## Debugging account registration failures

If a server has been updated to require `eab_policy=always-required` in the
[ACME configuration](/vault/api-docs/secret/pki#set-acme-configuration),
new account registration (and reuse of existing accounts will fail).

### Symptoms

> When registering a new account without an [External Account Binding
> (EAB)](/vault/api-docs/secret/pki#acme-external-account-bindings), the
> Vault Server will reject the request with a response like:
> 
> > Unable to register an account with ACME server
> 
> with further information provided in the debug logs (in the case of
> `certbot`):
> 
> > Server requires external account binding.
> 
> or, if the client incorrectly contacted the server, an error like:
> 
> > The request must include a value for the 'externalAccountBinding' field
> 
> In either case, a new account needs to be created with an EAB token created
> by Vault.

Reviewer: ```suggestion
Your Vault server has `eab_policy=always-required` and you see the error:

> Unable to register an account with ACME server

when registering a new account without an [External Account Binding
(EAB)](/vault/api-docs/secret/pki#acme-external-account-bindings).

You may also see errors similar to the following in your ACME debug logs:

- `Server requires external account binding.`
- `The request must include a value for the 'externalAccountBinding' field`

### Cause

If your [ACME configuration](/vault/api-docs/secret/pki#set-acme-configuration)
includes `eab_policy=always-required`, new account registration and reusing
existing accounts fail unless you provide an EAB token created by Vault.
```

Style issue: active voice

### Resolution

Using a Vault token, [fetch a new external account
binding](/vault/api-docs/secret/pki#get-acme-eab-binding-token) for
the [desired directory](/vault/api-docs/v1.14.x/secret/pki#acme-directories):

```
$ vault write -f pki/roles/my-role-name/acme/new-eab
...
directory roles/my-role-name/acme/directory
id        bc8088d9-3816-5177-ae8e-d8393265f7dd
key       MHcCAQE... additional data elided ...
...
```

Then pass these fields into the ACME client. For example, with `certbot`:

```
$ certbot [... additional parameters ...] \
    --server https://cluster-b.vault.example.com/v1/pki/roles/my-role-name/acme/directory \
    --eab-kid bc8088d9-3816-5177-ae8e-d8393265f7dd \
    --eab-hmac-key MHcCAQE... additional data elided ...
```

Ensure that the ACME directory passed to the ACME client matches that
fetched from the Vault.

## Debugging EAB directory bindings

> If an EAB account token is incorrectly used with the wrong directory, the
> ACME server will reject the request with an error about insufficient
> permissions.

Reviewer: ```suggestion
```

### Symptoms

> When initializing a new account against this Vault server, an error might
> appear like:
> 
> > The client lacks sufficient authorization :: failed to verify eab
> 
> This is caused by requesting an EAB from a directory not matching the
> one the client used.

Reviewer: ```suggestion
When initializing a new account against your Vault server, you see the error:

> The client lacks sufficient authorization :: failed to verify eab

### Cause

If you request an EAB account token from a directory different from the
directory used by the ACME client, the ACME server rejects the request because
the client lacks sufficient permission to access the token.
```
Style issue: active voice
Note: I'm not sure I fully understood which client/server is involved with each step so the suggested re-write my not be technically accurate.

### Resolution

Ensure the requested EAB token matches the directory. For a given directory
at `/some/path/acme/directory`, fetch EAB tokens from
`/some/path/amce/new-eab`. The remaining resolution steps are the same as
for [debugging account registration
failures](#debugging-account-registration-failures).

> ## Debugging issuance failures

Reviewer: ```suggestion
## Error: ACME validation failed for {CERTIFICATE_ID}
```

> When Vault can't verify the server's identity through the client's requested
> [challenge type](/vault/api-docs/secret/pki#acme-challenge-types) (`dns-01`,
> `http-01`, or `tls-alpn-01`), Vault will not issue the certificate requested
> by the client.

Reviewer: ```suggestion
```

### Symptoms

> In the Vault logs or returned by the client, an error might occur like:
> 
> > ACME validation failed for a465a798-4400-6c17-6735-e1b38c23de38-tls-alpn-01: ...
> 
> this indicates that the server was unable to validate this challenge accepted
> by the client.

Reviewer: ```suggestion
You see an error like:

> ACME validation failed for a465a798-4400-6c17-6735-e1b38c23de38-tls-alpn-01: ...

In the ACME client or when reviewing Vault logs.

### Cause

Validation errors indicate that the Vault server could not validate a challenge
that was accepted by the ACME client.

Vault will not issue certificates when the server identity cannot be verified
through a client [challenge](/vault/api-docs/secret/pki#acme-challenge-types)
request.
```

style issue: active voice

### Resolution

Ensure that DNS is configured correctly from the Vault server's perspective,
including setting [any custom DNS resolver](/vault/api-docs/secret/pki#dns_resolver).

Ensure that any firewalls are set up to allow Vault to talk to the relevant
systems (the DNS server in the case of `dns-01`, port 80 on the target
machine for `http-01`, or port 443 on the target machine for `tls-alpn-01`
challenges).

> ## Debugging account revocation

Reviewer: ```suggestion
## Error: The client lacks sufficient authorization: account in status: revoked
```

> Vault, via the [tidy mechanics](/vault/api-docs/secret/pki#tidy_acme), will
> periodically remove stale accounts.

Reviewer: ```suggestion
```

### Symptoms

> When Vault has removed an account that it has deemed stale, if a client
> attempts to reuse this account, an error will occur such as:
> 
> > The client lacks sufficient authorization: account in status: revoked

Reviewer: ```suggestion
You see the error:

> The client lacks sufficient authorization: account in status: revoked

when the ACME client tries to run.

### Cause

If you have [tidy mechanics](/vault/api-docs/secret/pki#tidy_acme) enabled,
Vault periodically removes stale accounts and rejects connections from ACME
client that try to reuse a revoked account.
```

### Resolution

Refer to the ACME client's documentation for removing cached local
configuration and setup a new account, specifying any EABs as required.

## Tutorial

Refer to the [Build Your Own Certificate Authority (CA)](/vault/tutorials/secrets-management/pki-engine)
guide for a step-by-step tutorial.

Have a look at the [PKI Secrets Engine with Managed Keys](/vault/tutorials/enterprise/managed-key-pki)

> for more about how to use externally managed keys with PKI.

Reviewer: Why is a general note about PKI tutorials relevant to a troubleshooting doc? What problem are you trying to solve by including this section?

Author: Same as above: https://github.com/hashicorp/vault/pull/21381#discussion_r1237609561 -- this is a common "footer" included in most every page about most every secrets engine/...

Reviewer: Again, that's really weird. Generally, boilerplate footers should be separate from page content so (1) readers don't think there's additional information in those sections that's **specifically** relevant to what they're reading, and (2) the footer text isn't considered/weighed in web search results.

## API

The PKI secrets engine has a full HTTP API. Please see the
[PKI secrets engine API](/vault/api-docs/secret/pki) for more

> details.

Reviewer: Same questions as above:
Why is a note about the API relevant to a troubleshooting doc? What problem are you trying to solve by including this section?

Author: This is a common footer included in all docs pages just about.
